app/models/assignment_generator.rb

```diff
@@ -0,0 +1,308 @@
+    @recipients_by_priority = @potential_givers.keys.sort_by do |key|
+      [@potential_givers[key].size, rand]
+    end
+  end
+
+  # Set up the hashtables we use to store our current assignments.
+  def set_up_assignments
+    # This table will contain a mapping from each request to the offer
+    # that they've been (temporarily) assigned to.
+    @assignment_as_recipient = {}
+
+    # This table will contain a mapping from each offer to the request
+    # that they've been (temporarily) assigned to.
+    @assignment_as_giver = {}
+  end
+
+  # Try to "improve" the assignments by looking for augmenting paths.
+  def improve_assignments
+    distance = augmenting_path_layers
+    return false if distance.nil?
+
+    # If we've reached this point, there is at least one augmenting path. It's
+    # time to use depth-first search to find as many as possible.
+    @recipients_by_priority.each do |recipient|
+      # We try to find a giver for each unassigned recipient.
+      next if @assignment_as_recipient[recipient]
+
+      find_giver_for_recipient(distance, recipient)
+    end
+
+    true
+  end
+
+  # Double check that the @assignment_as_giver and @assignment_as_recipient
+  # tables are consistent with each other, and that each recipient/giver pair is
+  # compatible (listed in the @potential_givers table). This should be
+  # guaranteed by the algorithm, but it's nice to know that it's not going to go
+  # catastrophically wrong.
+  def validate_assignments
+    @assignment_as_recipient.each_pair do |recipient, giver|
+      return false unless @assignment_as_giver[giver] == recipient
+      return false unless @potential_givers[recipient].include?(giver)
+    end
+
+    @assignment_as_giver.each_pair do |giver, recipient|
+      return false unless @assignment_as_recipient[recipient] == giver
+    end
+
+    true
+  end
+
+  # Save the assignments we calculated to the database.
+  def save_assignments
+    # First, build assignments for all of the signups that have been matched.
+    ChallengeAssignment.transaction do
+      @assignment_as_recipient.each_pair do |recipient, giver|
+        # This should never happen, but just in case:
+        next if recipient.nil? || giver.nil?
+
+        ChallengeAssignment.create(collection_id: @collection.id,
+                                   request_signup_id: recipient,
+                                   offer_signup_id: giver)
+      end
+    end
+
+    # Update the challenge_signups to indicate whether they've been assigned.
+    #
+    # TODO: These fields are no longer used for anything. Can the columns be
+    # dropped from the ChallengeSignup table?
+    ChallengeSignup.where(id: @assignment_as_recipient.keys)
+      .update_all(assigned_as_request: true)
+    ChallengeSignup.where(id: @assignment_as_recipient.values)
+      .update_all(assigned_as_offer: true)
+  end
+
+  # Add blank assignments for any users that couldn't be assigned.
+  def save_placeholder_assignments
```

Comment:
Why are placeholder assignments saved in a separate transaction? (I don't have a reason for or against, though.)

Reply:
No particular reason. I think I originally split them into different functions to keep the code more organized, and as a side effect split the transaction as well. But I can move the transaction block into `generate` to make it a single transaction.

Reply:
You have it in one transaction now, still looks fine to me. How long did the single transaction take in your 2k participants test?

Reply:
On my machine, the transaction takes ~6 seconds (out of ~9 seconds total for regenerating assignments). I'm not sure how much time that would translate into in production, but hopefully it'll be fast enough to prevent most locking issues.